File: java/src/com/recursion/inout/Letter.java
package com.recursion.inout;

import java.util.ArrayList;
import java.util.List;

public class Letter {

    public void printLetterCaseChange(String in) {
        String out = "";
        List<String> list = new ArrayList<>();
        generate(in, out, list);
        System.out.println(list);
    }

    private void generate(String in, String out, List<String> list) {
        if (in.length() == 0) {
            list.add(out);
            return;
        }
        if (Character.isDigit(in.charAt(0))) {
            String out1 = out;
            out1 = out1.concat(String.valueOf(in.charAt(0)));
            in = in.substring(1);
            generate(in, out1, list);
        } else {
            String op1 = out;
            String op2 = out;
            op1 = op1.concat(String.valueOf(in.charAt(0)).toLowerCase());
            op2 = op2.concat(String.valueOf(in.charAt(0)).toUpperCase());
            in = in.substring(1);
            generate(in, op1, list);
            generate(in, op2, list);
        }
    }
}

File: js/recursion/inout/letter.js
export const printLetterCaseChange = (inp) => {
  let out = "";
  const arr = [];
  generate(inp, out, arr);
  console.log(arr);
};

const generate = (inp, out, arr) => {
  if (inp.length == 0) {
    arr.push(out);
    return;
  }
  if (isDigit(inp[0])) {
    let out1 = out;
    out1 = out1 + inp[0];
    inp = inp.substring(1);
    generate(inp, out1, arr);
  } else {
    let out1 = out;
    let out2 = out;
    out1 = out1 + inp[0].toUpperCase();
    out2 = out2 + inp[0].toLowerCase();
    inp = inp.substring(1);
    generate(inp, out1, arr);
    generate(inp, out2, arr);
  }
};

const isDigit = (x) => {
  return !isNaN(x - parseInt(x));
};
